File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.length() < t.length()){
            return "";
        }
        unordered_map<char, int> map;
        for(int i=0; i<t.length(); i++){
            map[t[i]]++;
        }
        int count=0, start=0, left=0, right=0, min_length = INT_MAX;
        while ( right < s.length() ) {
            if (map[s[right]] > 0) {
                count++;
            }
            map[s[right]]--;
            if (count == t.length()) {
                while (left < right && map[s[left]]<0) {
                    map[s[left]]++;
                    left++;
                }
                if (min_length > (right - left)){
                    min_length = right - left + 1;
                    start = left;
                }
                map[s[left]]++;
                left++;
                count--;
            }
            right++;
        }
        if (min_length > s.length()){
            return "";
        } else {
            return s.substr(start, min_length);
        }

    }
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""
        
        window = {}

        for ch in t:
            window[ch] = 1 + window.get(ch, 0)
        
        left, count, start, min_len = 0, 0, 0, float("infinity")

        for right in range(len(s)):
            ch = s[right]

            if window.get(ch, 0) > 0:
                count += 1
            
            window[ch] = window.get(ch, 0) - 1

            if count == len(t):
                ch2 = s[left]

                while left < right and window.get(ch2, 0) < 0:
                    window[ch2] = 1 + window.get(ch2, 0)
                    left += 1
                    ch2 = s[left]
                
                if min_len > (right - left):
                    min_len = right -left + 1
                    start = left
                
                window[ch2] = 1 + window.get(ch2, 0)
                left += 1
                count -= 1
        
        return s[start:start+min_len] if min_len <= len(s) else ""
